### PR TITLE
add crl.get_revoked_certificate method

### LINKED ISSUE #4330: Add method to check if cert is in CRL?
Hi,

I am using `cryptography` to check certificates against CRLs, to see if they are revoked, and I find the current CRL implementation a little lacking. To check if a certificate is revoked, I have to retrieve all the `RevokedCertificate` from the `CertificateRevocationList` and see if any of them has the same serial as the certificate in question (if there's a better way, please let me know!). This is time consuming on big CRLs.

I have played around with `X509_CRL_get0_by_serial()` and `X509_CRL_get0_by_cert()` from OpenSSL, and have found them to drastically improve performance.

So, what do you think about adding a function to `CertificateRevocationList` that checks if a specified certificate is in the CRL? I'm thinking `is_certificate_revoked` or `get_revoked_certificate` which can return a `RevokedCertificate`.

Thanks for making cryptography!

### PR DESCRIPTION
**Pull request by reaperhulk**
fixes #4330

### CODE REVIEW

File: docs/x509/reference.rst [hunk 1]
```diff
@@ -463,6 +463,15 @@ X.509 CRL (Certificate Revocation List) Object
             >>> crl.fingerprint(hashes.SHA256())
             b'e\xcf.\xc4:\x83?1\xdc\xf3\xfc\x95\xd7\xb3\x87\xb3\x8e\xf8\xb93!\x87\x07\x9d\x1b\xb4!\xb9\xe4W\xf4\x1f'
 
+    .. method:: get_revoked_certificate(serial_number)
```
Comment (alex) [inline docs/x509/reference.rst]: I think this should be `get_revoked_certificate_by_serial_number`?

File: src/cryptography/hazmat/backends/openssl/x509.py [hunk 2]
```diff
@@ -235,6 +238,22 @@ def fingerprint(self, algorithm):
         h.update(der)
         return h.finalize()
 
+    def get_revoked_certificate(self, serial_number):
+        revoked = self._backend._ffi.new("X509_REVOKED **")
+        asn1_int = _encode_asn1_int_gc(self._backend, serial_number)
+        res = self._backend._lib.X509_CRL_get0_by_serial(
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/x509.py]: Does this do something more efficient than just a linear search?

File: src/cryptography/hazmat/backends/openssl/x509.py [hunk 3]
```diff
@@ -235,6 +238,22 @@ def fingerprint(self, algorithm):
...
+            self._x509_crl, revoked, asn1_int
+        )
+        if res == 0:
+            return None
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/x509.py]: Does this push an error? Do we need to clear the stack?

File: src/cryptography/hazmat/backends/openssl/x509.py (hunk 2 above)
Comment (reaperhulk) [reply src/cryptography/hazmat/backends/openssl/x509.py]: Implementation here: https://github.com/openssl/openssl/blob/01ad66f85d22fd001582b5f2e6e18db8b820c550/crypto/x509/x_crl.c#L381-L414

Short answer: not really. It sorts and then does a linear asn1 integer comparison. But no copies across ffi boundary.

Comment (reaperhulk) [reply src/cryptography/hazmat/backends/openssl/x509.py]: Ah, actually the stack find uses a binary search once it sorts the list. So it is faster than a linear search (in most cases)

File: src/cryptography/hazmat/backends/openssl/x509.py (hunk 3 above)
Comment (reaperhulk) [reply src/cryptography/hazmat/backends/openssl/x509.py]: No errors are pushed onto the stack in the test cases we have, but you'll never get me to say I'm 100% confident there's no scenario under which OpenSSL won't push an error onto it from some other function it calls.

File: docs/x509/reference.rst (hunk 1 above)
Comment (reaperhulk) [reply docs/x509/reference.rst]: Yeah that's more consistent with `get_values_for_type`, `get_extension_for_oid`, `get_extension_for_class`. Changing.

**Pull request merged**